File: C++/1.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <map>
#include <algorithm>

using namespace std;

void number11() {
    srand(time(0));

    int N;
    cout << "Задача 1: \nВведите N: ";
    cin >> N;

    vector<vector<int>> matrix(N, vector<int>(N));

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            matrix[i][j] = rand() % 201 - 100;
        }
    }

    int maxPosCount = -1;
    int bestRowIdx = -1;

    for (int i = 0; i < N; ++i) {
        int posCount = 0;
        for (int j = 0; j < N; ++j) {
            if (matrix[i][j] > 0) ++posCount;
        }
        if (posCount > maxPosCount) {
            maxPosCount = posCount;
            bestRowIdx = i;
        }
    }

    vector<int> resultRow;
    if (bestRowIdx != -1) {
        resultRow = matrix[bestRowIdx];
    }

    cout << "\nМатрица:\n";
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cout.width(4);
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }

    cout << "\nСтрока с наибольшим количеством положительных чисел ("
        << maxPosCount << "): ";
    for (int x : resultRow) {
        cout << x << " ";
    }
    cout << endl;
    cout << endl;
}


void number12() {
    srand(static_cast<unsigned>(time(0)));

    int M, N;
    cout << "Задача 2\nВведите M и N: ";
    cin >> M >> N;

    vector<vector<int>> matrix(M, vector<int>(N));
    map<int, int> freq;

    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            matrix[i][j] = 100 + rand() % 51;
            freq[matrix[i][j]]++;
        }
    }

    int maxFreq = 0;
    int mostFreqNum = -1;
    for (const auto& p : freq) {
        if (p.second > maxFreq) {
            maxFreq = p.second;
            mostFreqNum = p.first;
        }
    }

    bool allUnique = true;
    for (const auto& p : freq) {
        if (p.second > 1) {
            allUnique = false;
            break;
        }
    }

    cout << "\nМатрица:\n";
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            cout << matrix[i][j] << " ";
        }
        cout << "\n";
    }

    if (allUnique) {
        cout << "\nВсе числа уникальны.\n\n";
    }
    else {
        cout << "\nНаиболее встречающееся число: " << mostFreqNum
            << " (встречается " << maxFreq << " раз)\n\n";
    }
}


void clearScreen() {
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

int countNeighbors(const vector<vector<int>>& grid, int i, int j, int rows, int cols) {
    int count = 0;
    for (int di = -1; di <= 1; di++) {
        for (int dj = -1; dj <= 1; dj++) {
            if (di == 0 && dj == 0) continue;
            int ni = (i + di + rows) % rows;
            int nj = (j + dj + cols) % cols;
            count += grid[ni][nj];
        }
    }
    return count;
}

vector<vector<int>> nextGen(const vector<vector<int>>& grid, int mode) {
    int rows = grid.size(), cols = grid[0].size();
    vector<vector<int>> next(rows, vector<int>(cols, 0));

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            int alive = grid[i][j];
            int n = countNeighbors(grid, i, j, rows, cols);

            if (mode == 0) {
                if (alive) {
                    if (n == 2 || n == 3) next[i][j] = 1;
                }
                else {
                    if (n == 3) next[i][j] = 1;
                }
            }
            else {
                if (alive) {
                    if (n == 1 || n == 3 || n == 5) next[i][j] = 1;
                }
                else {
                    if (n == 2 || n == 4) next[i][j] = 1;
                }
            }
        }
    }
    return next;
}

void placeGlider(vector<vector<int>>& grid, int r, int c) {
    int rows = grid.size(), cols = grid[0].size();
    grid[(r + 0) % rows][(c + 1) % cols] = 1;
    grid[(r + 1) % rows][(c + 2) % cols] = 1;
    grid[(r + 2) % rows][(c + 0) % cols] = 1;
    grid[(r + 2) % rows][(c + 1) % cols] = 1;
    grid[(r + 2) % rows][(c + 2) % cols] = 1;
}

void randomInit(vector<vector<int>>& grid, double density = 0.3) {
    int rows = grid.size(), cols = grid[0].size();
    for (int i = 0; i < rows; ++i)
        for (int j = 0; j < cols; ++j)
            grid[i][j] = (rand() / double(RAND_MAX)) < density ? 1 : 0;
}


void number13() {
    srand(static_cast<unsigned>(time(0)));

    int rows = 20, cols = 50;
    int mode = 0;
    int steps = 200;

    cout << "Задача 3:\n";
    cout << "Выберите режим:\n0 - классические правила\n1 - альтерантивные\n";
    cin >> mode;

    vector<vector<int>> grid(rows, vector<int>(cols, 0));

    cout << "Инициализация:\n1 - глайдер\n2 - случайные\n";
    int init;
    cin >> init;

    if (init == 1) {
        placeGlider(grid, rows / 2 - 1, cols / 2 - 2);
    }
    else if (init == 2) {
        randomInit(grid, 0.25);
    }
    else {
        cout << "Неверный выбор. По умолчанию случ\n";
        randomInit(grid, 0.2);
    }

    cout << "Симуляция\n";
    for (int gen = 0; gen < steps; gen++) {
        clearScreen();
        cout << "Поколение: " << gen << " | Режим: " << (mode ? "альтернативный" : "классический") << "\n";
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                cout << (grid[i][j] ? '0' : ' ');
            }
            cout << '\n';
        }
        grid = nextGen(grid, mode);
        
    }

    cout << "\nСимуляция завершена.\n\n";
}

File: Go/1.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)


func countNeighbors(grid [][]int, i, j, rows, cols int) int {
	count := 0
	for di := -1; di <= 1; di++ {
		for dj := -1; dj <= 1; dj++ {
			if di == 0 && dj == 0 {
				continue
			}
			ni := (i + di + rows) % rows
			nj := (j + dj + cols) % cols
			count += grid[ni][nj]
		}
	}
	return count
}

func nextGen(grid [][]int, mode int) [][]int {
	rows := len(grid)
	cols := len(grid[0])
	next := make([][]int, rows)
	for i := range next {
		next[i] = make([]int, cols)
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			alive := grid[i][j]
			n := countNeighbors(grid, i, j, rows, cols)
			if mode == 0 {
				if alive == 1 {
					if n == 2 || n == 3 {
						next[i][j] = 1
					}
				} else {
					if n == 3 {
						next[i][j] = 1
					}
				}
			} else {
				if alive == 1 {
					if n == 1 || n == 3 || n == 5 {
						next[i][j] = 1
					}
				} else {
					if n == 2 || n == 4 {
						next[i][j] = 1
					}
				}
			}
		}
	}
	return next
}

func placeGlider(grid [][]int, r, c int) {
	rows := len(grid)
	cols := len(grid[0])
	grid[(r+0)%rows][(c+1)%cols] = 1
	grid[(r+1)%rows][(c+2)%cols] = 1
	grid[(r+2)%rows][(c+0)%cols] = 1
	grid[(r+2)%rows][(c+1)%cols] = 1
	grid[(r+2)%rows][(c+2)%cols] = 1
}

func randomInit(grid [][]int, density float64) {
	rows := len(grid)
	cols := len(grid[0])
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if rand.Float64() < density {
				grid[i][j] = 1
			} else {
				grid[i][j] = 0
			}
		}
	}
}

func number11() {
	rand.Seed(time.Now().UnixNano())
	var N int
	fmt.Print("Задача 1: \nВведите N: ")
	fmt.Scanln(&N)

	matrix := make([][]int, N)
	for i := range matrix {
		matrix[i] = make([]int, N)
		for j := range matrix[i] {
			matrix[i][j] = rand.Intn(201) - 100
		}
	}

	maxPosCount := -1
	bestRowIdx := -1
	for i := 0; i < N; i++ {
		posCount := 0
		for j := 0; j < N; j++ {
			if matrix[i][j] > 0 {
				posCount++
			}
		}
		if posCount > maxPosCount {
			maxPosCount = posCount
			bestRowIdx = i
		}
	}

	var resultRow []int
	if bestRowIdx != -1 {
		resultRow = make([]int, N)
		copy(resultRow, matrix[bestRowIdx])
	}

	fmt.Println("\nМатрица:")
	for _, row := range matrix {
		for _, val := range row {
			fmt.Printf("%4d ", val)
		}
		fmt.Println()
	}

	fmt.Printf("\nСтрока с наибольшим количеством положительных чисел (%d): ", maxPosCount)
	for _, x := range resultRow {
		fmt.Print(x, " ")
	}
	fmt.Println()
	fmt.Println()
}

func number12() {
	rand.Seed(time.Now().UnixNano())
	var M, N int
	fmt.Print("Задача 2\nВведите M и N: ")
	fmt.Scanln(&M, &N)

	matrix := make([][]int, M)
	freq := make(map[int]int)
	for i := range matrix {
		matrix[i] = make([]int, N)
		for j := range matrix[i] {
			matrix[i][j] = 100 + rand.Intn(51)
			freq[matrix[i][j]]++
		}
	}

	maxFreq := 0
	mostFreqNum := -1
	for num, cnt := range freq {
		if cnt > maxFreq {
			maxFreq = cnt
			mostFreqNum = num
		}
	}

	allUnique := true
	for _, cnt := range freq {
		if cnt > 1 {
			allUnique = false
			break
		}
	}

	fmt.Println("\nМатрица:")
	for _, row := range matrix {
		for _, val := range row {
			fmt.Print(val, " ")
		}
		fmt.Println()
	}

	if allUnique {
		fmt.Println("\nВсе числа уникальны — нет числа, встречающегося чаще других.")
	} else {
		fmt.Printf("\nНаиболее часто встречающееся число: %d (встречается %d раз(а))\n", mostFreqNum, maxFreq)
	}
}

func number13() {
	rand.Seed(time.Now().UnixNano())
	rows, cols := 20, 50
	mode := 0
	steps := 200

	fmt.Print("Задача 3:\n")
	fmt.Print("Выберите режим:\n0 — классические правила\n1 — альтернативные\n")
	fmt.Scanln(&mode)

	grid := make([][]int, rows)
	for i := range grid {
		grid[i] = make([]int, cols)
	}

	fmt.Print("Инициализация:\n1 — глайдер\n2 — случайная колония\n")
	var init int
	fmt.Scanln(&init)

	switch init {
	case 1:
		placeGlider(grid, rows/2-1, cols/2-2)
	case 2:
		randomInit(grid, 0.25)
	default:
		fmt.Println("Неверный выбор. Используется случайная инициализация (20%).")
		randomInit(grid, 0.2)
	}

	fmt.Println("Cимуляция")
	for gen := 0; gen < steps; gen++ {
		modeStr := "классический"
		if mode != 0 {
			modeStr = "альтернативный"
		}
		fmt.Printf("Поколение: %d | Режим: %s\n", gen, modeStr)
		for i := 0; i < rows; i++ {
			for j := 0; j < cols; j++ {
				if grid[i][j] == 1 {
					fmt.Print("0")
				} else {
					fmt.Print(" ")
				}
			}
			fmt.Println()
		}
		grid = nextGen(grid, mode)
	}
	fmt.Println("\nСимуляция завершена.")
}
